actions/sync_gws_mailing_lists.py

```diff
@@ -0,0 +1,125 @@
+Keycloak mailing list groups. Users are subscribed to Google Workspace groups
+using their KeyCloak `canonical_email` attribute.
+
+Keycloak mailing list groups are the subgroups of /mail. Each group must
+have attribute `email` that will be used to map it to a Google Workspace
+group.
+
+KeyCloak client is configured using environment variables. See krs/token.py
+for details.
+
+Domain-wide delegation must be enabled for the service account. See code
+for which scopes are required. Admin API must be enabled in the Google
+Cloud Console project.
+
+The delegator principal must have the Google Workspace admin role, have
+access to the service account, and have "Service Account Token Creator"
+role.
+"""
+import asyncio
+import logging
+
+from googleapiclient.discovery import build
+from google.oauth2 import service_account
+
+from krs.token import get_rest_client
+from krs.groups import get_group_membership, group_info
+from krs.users import user_info
+
+
+logger = logging.getLogger('sync_gws_mailing_lists')
+
+
+def get_all_group_members(group_email, gws_members_client):
+    """Get email addresses of all group members"""
+    req = gws_members_client.list(groupKey=group_email)
+    members = []
+    while req is not None:
+        res = req.execute()
+        if 'members' in res:
+            members.extend(m['email'] for m in res['members'])
+        else:
+            break
+        req = gws_members_client.list_next(req, res)
+    return members
+
+
+async def sync_gws_mailing_lists(gws_members_client, keycloak_client, dryrun=False):
+    """Synchronize memberships of Google Workspace groups to their corresponding
+    Keycloak mailing list groups.
+
+    Keycloak mailing list groups are the subgroups of /mail. Each group must
+    have attribute `email` that will be used to map it to a Google Workspace
+    group.
+
+    Args:
+        gws_members_client (googleapiclient.discovery.Resource): Directory API's Members resource
+        keycloak_client (OpenIDRestClient): REST client to the KeyCloak server
+
+    Returns:
+        dict: run summary for unit tests
+    """
+    debug = {"added": set(), "deleted": set()}
+    ml_root_group = await group_info('/mail', keycloak_client)
+    ml_groups = [sg for sg in ml_root_group['subGroups'] if sg['name'] != '_admin']
+    for ml_group in ml_groups:
+        if 'email' not in ml_group['attributes']:
+            logger.error(f"Group {ml_group['path']} doesn't have attribute 'email'")
+            continue
+        group_email = ml_group['attributes']['email'][0]
+        kc_usernames = await get_group_membership(ml_group['path'], rest_client=keycloak_client)
+        kc_users = [await user_info(u, rest_client=keycloak_client) for u in kc_usernames]
+        kc_emails = [u['attributes']['canonical_email'] for u in kc_users]
+        gws_emails = get_all_group_members(group_email, gws_members_client)
+
+        missing_emails = [e for e in kc_emails if e not in gws_emails]
```

Comment:
This works fine.  But it feels like it should be a set operation `kc_emails - gws_emails`.  Same for the `unwanted_emails`.